### ISSUE TITLE
Add average-case robustness adapter via imagecorruptions (closes Ch.2 scope gap)

### ISSUE THREAD

**Issue by stanlrt** [CAIIVS/raitap#213]
## Motivation

Current robustness coverage in `raitap` is **worst-case adversarial only**:
- Empirical attacks (`torchattacks`, `foolbox`) — verdict in {ATTACKED, NOT_ATTACKED}
- Formal verification (`marabou`, `alpha-beta-CROWN`, `auto-LiRPA`) — verdict in {VERIFIED, FALSIFIED, UNKNOWN}

All target `forall x' in B_eps(x): f(x') = f(x)`.

The thesis Chapter 2 currently declares **average-case (natural / environmental) robustness** out of scope. This issue proposes closing that gap with a dedicated adapter so the contract argument in the thesis covers both worst-case AND average-case robustness, matching the title ("Algorithms for the assessment of trustworthiness of AI-based systems").

Average-case verdict shape is fundamentally different: instead of a per-input boolean (forall-quantifier, single counterexample falsifies), we report `P_{x' ~ B(x)}[f(x') = f(x)] >= 1 - delta` — accuracy with confidence interval under a perturbation distribution (Gaussian noise, fog, brightness, JPEG compression, ...). This aligns with the confidence-interval tests demonstrated by Denzel et al. (2024, Table III).

## Conceptual framing — worst-case vs average-case

| | Worst-case (existing) | Average-case (this issue) |
|---|---|---|
| Quantifier | forall x' in B(x) | E_{x'~p} or Pr[.] >= 1 - delta |
| B(x) shape | norm ball, small eps | distribution, realistic noise levels |
| Granularity | per-input | per-dataset (inherently aggregate) |
| Falsification | single counterexample | statistic over n samples |
| Threat model | adversary picks worst input in ball | environment samples from distribution |
| Pipeline call | `A: (f, x, B(x), t) -> v` | `A_avg: (f, D, B, t) -> v` (takes dataset, not single input) |

Worst-case implies average-case trivially; converse false. They answer different questions, not subsets of each other.

## Proposed library

[`imagecorruptions`](https://github.com/bethgelab/imagecorruptions) (PyPI: `imagecorruptions`, v1.1.2, Apache-2.0).

- Canonical reference implementation of the Hendrycks and Dietterich ImageNet-C corruption suite: 15 corruptions + 4 validation corruptions, 5 severity levels each.
- Tiny API surface (`corrupt(img, corruption_name=..., severity=...)`); works on numpy arrays / PIL images. Generalises to arbitrary image dims, aspect ratios, and grayscale.
- Apache-2.0.
- Acceptable staleness: last release 2020-12. The corruption set is **definitionally frozen** by the benchmark — the package functions as the reference implementation of an established standard, so no API churn is expected or wanted.

### Dependency footprint — verified

From `setup.py` (v1.1.2):
```
numpy >= 1.16
Pillow >= 5.4.1
scikit-image >= 0.15
opencv-python >= 3.4.5
scipy >= 1.2.1
numba >= 0.53.0
```

- **No torch.** Operates on numpy / PIL images; will not fight raitap's torch pin.
- **No Wand / no system ImageMagick.** v1.1.2 explicitly removed the `libmagickwand` dependency and reimplemented corruptions in Python. `uv run` works out of the box.
- All lower-bounded, no upper caps → modern resolver should be happy.
- **Sole modest risk:** numba vs numpy version alignment. numba lags numpy ABI updates. If raitap pins numpy >= 2.0, ensure `numba >= 0.60`. Likely a non-issue with current uv lock; verify during step 0 below.

### Fallback if dep resolution fails

If for any reason `uv add imagecorruptions` cannot resolve in raitap's env, **vendor the 15 corruption functions** into `raitap/robustness/avg_case/_corruptions.py`:
- ~500 LOC, pure Python, Apache-2.0 → permits vendoring with attribution
- Eliminates external dep entirely; only inherits raitap's existing numpy/scipy/scikit-image/opencv-python
- Keep the upstream LICENSE header and a `_VENDORED_FROM` constant pointing at the source commit

This fallback is acceptable for a BA-scope project; the corruption set is frozen so vendored code does not bit-rot.

### Alternatives considered

- **robustbench** — active and newer, but heavy and dataset-coupled (loaders for CIFAR-10-C / ImageNet-3DCC, model zoo). We do not want to inherit its dataset/model-zoo surface; raitap should be able to perturb arbitrary user-provided tensors.
- **albumentations / augly / torchvision.v2.transforms** — general augmentation, not the standardised corruption benchmark; would re-invent the H&D corruption vocabulary and lose ImageNet-C parity.
- **IBM ART** — covers many threat classes (evasion, poisoning, extraction, inference) but its common-corruption support is secondary. Adversarial coverage would heavily overlap with foolbox/torchattacks already in raitap. Reserved as the natural pick if raitap later extends to Denzel's security pillar (poisoning / extraction / inference).

## Architecture sketch

The existing two-pillar adapter pattern in `raitap/src/raitap/robustness/` extends naturally with a third method kind.

### Contract changes (`raitap/robustness/contracts.py`)
- Add `MethodKind.AVERAGE_CASE` alongside `EMPIRICAL_ATTACK`, `FORMAL_VERIFICATION`.
- Add `PerturbationDistribution` dataclass (sibling of `PerturbationBudget`):
  ```python
  @dataclass(frozen=True)
  class PerturbationDistribution:
      corruption_name: str        # one of the 15 H&D corruptions
      severity: int               # 1..5
      n_samples_per_input: int    # MC draws per input; 1 is fine since H&D corruptions are mostly deterministic
  ```
- Widen `RobustnessSemantics.budget: PerturbationBudget` to a union `PerturbationBudget | PerturbationDistribution`, OR add an optional `distribution` field — pick whichever keeps existing worst-case consumers unchanged.
- Add `StatisticalOutcome` dataclass (sibling of `VerificationOutcome`):
  ```python
  @dataclass(frozen=True)
  class StatisticalOutcome:
      n_samples: int
      n_correct: int
      accuracy: float
      ci_low: float               # 95% CI lower
      ci_high: float              # 95% CI upper
      clean_accuracy: float | None = None   # reference accuracy without perturbation
  ```
- **Verdict question:** decide whether to (a) keep verdict purely numerical (accuracy + CI, no enum) or (b) add `Verdict.ACCURATE` / `Verdict.DEGRADED` based on a user-supplied threshold. Recommendation: (a) — keep verdict numerical, let downstream reporting layer apply thresholds. Avoids an additional config knob in the contract.

### Assessor base (`raitap/robustness/assessors/base_assessor.py`)
- Add `AverageCaseAssessor(BaseAssessor)` analogous to `EmpiricalAttackAssessor` / `FormalVerificationAssessor`.
- Framework owns `assess(model, dataset, ...)`; subclasses implement `apply_perturbation(image, severity) -> image`.
- Note the signature change: takes a **dataset** (or iterable of inputs) rather than a single input. This is intrinsic to average-case — a single-input "average accuracy" is meaningless.

### Adapter (`raitap/robustness/assessors/imagecorruptions_assessor.py`)
- `@robustness_adapter(registry_name="imagecorruptions", library="imagecorruptions", algorithm_registry={...})`.
- `algorithm_registry` keyed by the 15 corruption names: `gaussian_noise`, `shot_noise`, `impulse_noise`, `defocus_blur`, `glass_blur`, `motion_blur`, `zoom_blur`, `snow`, `frost`, `fog`, `brightness`, `contrast`, `elastic_transform`, `pixelate`, `jpeg_compression` — each with `method_kind=AVERAGE_CASE`, `threat_model=BLACK_BOX_DECISION`.
- Use lazy import pattern (mirroring `torchattacks` / `foolbox`) so missing `imagecorruptions` does not break worst-case workflows.

### Semantics builder (`raitap/robustness/semantics.py`)
- Branch on `method_kind` to assemble either a `PerturbationBudget` (existing worst-case) or a `PerturbationDistribution` (new).

### Results (`raitap/robustness/results.py`)
- Aggregate per-sample correctness into a `RobustnessMetrics` extension carrying accuracy and CI (Wilson or Clopper-Pearson interval, default 95%).
- Per-input correctness booleans need not be stored — keep aggregate only, plus optional sample-level dump under a debug flag.

### Pipeline orchestrator
- Existing orchestrator iterates inputs and calls per-input assessors. For average-case, the orchestrator must hand the assessor a **dataset handle** (e.g. a DataLoader or iterable of tensors) rather than looping single inputs. Either:
  - (a) extend the orchestrator to detect `MethodKind.AVERAGE_CASE` and switch to dataset-handle mode, OR
  - (b) keep the per-input loop but have the average-case assessor accumulate internally and emit a single final verdict.
- Recommendation: (a) — cleaner contract, no hidden state in the assessor.

## Implementation order (for the impl agent)

1. **Step 0 — dep check.** `uv add imagecorruptions` in raitap; run existing test suite. If green, continue. If numpy/numba conflict, either bump pins or vendor (see fallback).
2. Add `MethodKind.AVERAGE_CASE`, `PerturbationDistribution`, `StatisticalOutcome` to `contracts.py`. Ensure existing worst-case serialisation unchanged. Add unit tests for serialisation of the new types.
3. Add `AverageCaseAssessor` base class. Define its `assess(model, dataset, distribution, target)` signature and the abstract `apply_perturbation`. Unit-test the aggregation logic with a dummy stub perturbation.
4. Add the `imagecorruptions_assessor.py` adapter. Lazy-import `imagecorruptions`. Register all 15 corruption names.
5. Extend the semantics builder + results module. Update `_serialisable_semantics` to handle the new distribution / outcome types.
6. Extend the orchestrator (option (a) above) to route `AVERAGE_CASE` requests with a dataset handle.
7. Demo end-to-end: one corruption (Gaussian noise) at one severity on a tiny torchvision classifier with CIFAR-10 / ISIC sample data. Verify the verdict serialises cleanly.

## Out of scope for this issue

- Certified / smoothing-based statistical robustness (e.g. randomised smoothing, Cohen et al. 2019) — different probabilistic guarantee shape; will be its own adapter if added.
- Loading pretrained `robustbench` model zoo entries.
- 3D / video / audio corruption suites.
- Training-time data augmentation use (the upstream library is explicit: corruptions are evaluation-only).
- Threshold-based ACCURATE / DEGRADED verdict — defer to a reporting layer.

## Acceptance criteria

- [ ] `MethodKind.AVERAGE_CASE`, `PerturbationDistribution`, `StatisticalOutcome` added to `raitap/robustness/contracts.py`; existing worst-case verdicts and serialisation unaffected.
- [ ] `AverageCaseAssessor` base class added in `base_assessor.py`.
- [ ] `imagecorruptions_assessor.py` adapter registered with all 15 corruption names; lazy import.
- [ ] Pipeline orchestrator routes `AVERAGE_CASE` calls with a dataset handle.
- [ ] One demo use case (Gaussian noise on a small classifier) end-to-end through the pipeline; produces accuracy + 95% CI.
- [ ] `RobustnessResult` from an average-case run serialises through the existing contract (JSON via `_serialisable_semantics`) without breaking transparency consumers.
- [ ] Unit tests in `raitap/robustness/tests/` cover: contract serialisation of new types, adapter registration, semantics builder for `AVERAGE_CASE`, end-to-end accuracy aggregation on a tiny tensor batch with a stub perturbation.
- [ ] `imagecorruptions` declared as an optional/extras dependency in `pyproject.toml` (mirroring the lazy-import pattern used for `torchattacks` / `foolbox`).
- [ ] If vendoring fallback was used: `_corruptions.py` carries upstream LICENSE header and `_VENDORED_FROM = "imagecorruptions vX.Y.Z @ <commit-sha>"`.
